File: frontend/src/types/campaign.ts
// Campaign types for M2 Multiplayer support

export type CampaignStatus = 'active' | 'paused' | 'ended' | 'archived'
export type CampaignRole = 'keeper' | 'co-keeper' | 'player' | 'observer'
export type MemberStatus = 'active' | 'inactive' | 'kicked' | 'left'

export interface Campaign {
  id: string
  name: string
  description: string | null
  keeper_id: number
  scenario_id: string | null
  invite_code: string
  max_players: number
  status: CampaignStatus
  settings: Record<string, any>
  created_at: string
  updated_at: string
}

export interface CampaignMember {
  id: string
  campaign_id: string
  user_id: number
  character_id: number | null
  role: CampaignRole
  status: MemberStatus
  joined_at: string
  last_seen_at: string | null
}

export interface CreateCampaignRequest {
  name: string
  description?: string
  max_players?: number
  scenario_id?: string
  settings?: Record<string, any>
}

export interface UpdateCampaignRequest {
  name?: string
  description?: string
  max_players?: number
  scenario_id?: string
  status?: CampaignStatus
  settings?: Record<string, any>
}

export interface JoinCampaignRequest {
  invite_code: string
  character_id?: number
}

export interface UpdateMemberRoleRequest {
  role: CampaignRole
}

export interface Member {
  id: string
  user_id: number
  username?: string
  character_id: number | null
  character_name?: string | null
  role: CampaignRole
  status: MemberStatus
  joined_at: string
  last_seen_at: string | null
}

File: backend/src/schemas/campaign.py
"""Campaign Pydantic schemas for request/response validation."""
from pydantic import BaseModel, Field, field_validator
from typing import Optional, Dict, Any


class CampaignCreate(BaseModel):
    """Schema for creating a new campaign."""
    name: str = Field(..., min_length=1, max_length=100, description="Campaign name")
    description: Optional[str] = Field(None, max_length=5000, description="Campaign description")
    max_players: Optional[int] = Field(4, ge=1, le=10, description="Maximum number of players")
    scenario_id: Optional[str] = Field(None, description="Associated scenario ID")
    settings: Optional[Dict[str, Any]] = Field(default_factory=dict, description="Campaign settings")


class CampaignUpdate(BaseModel):
    """Schema for updating a campaign."""
    name: Optional[str] = Field(None, min_length=1, max_length=100)
    description: Optional[str] = Field(None, max_length=5000)
    max_players: Optional[int] = Field(None, ge=1, le=10)
    scenario_id: Optional[str] = None
    status: Optional[str] = Field(None, pattern="^(active|paused|ended|archived)$")
    settings: Optional[Dict[str, Any]] = None


class CampaignJoin(BaseModel):
    """Schema for joining a campaign with invite code."""
    invite_code: str = Field(..., min_length=8, max_length=8, description="Invite code")
    character_id: Optional[int] = Field(None, description="Character ID to use")


class MemberRoleUpdate(BaseModel):
    """Schema for updating member role."""
    role: str = Field(..., pattern="^(keeper|co-keeper|player|observer)$")


class MemberAdd(BaseModel):
    """Schema for adding a member directly (keeper only)."""
    user_id: int = Field(..., description="User ID to add")
    role: str = Field(default="player", pattern="^(keeper|co-keeper|player|observer)$")
    character_id: Optional[int] = Field(None, description="Character ID")
